subsys/net/lib/coap/coap.c

```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 		return NULL;
 	}
 
+	type = coap_header_get_type(response);
 	id = coap_header_get_id(response);
 	tkl = coap_header_get_token(response, token);
 
+	if (type == COAP_TYPE_ACK || type == COAP_TYPE_RESET) {
+		piggybacked = true;
+	}
+
 	for (i = 0, r = replies; i < len; i++, r++) {
 		int age;
 
 		if ((r->id == 0U) && (r->tkl == 0U)) {
```

Comment:
Is the message ID 0 used to indicate something special in the zephyr implementation?
if not, what is the purpose of this check?

FMPOV, it's wrong. I don't see this is related to RFC 7252. 
Considering a separate response with empty token, assumptions about the used MIDs will fail. The other implementation may use 0 without considering any special processing, as I don't see one in RFC 7252

Reply:
Can't tell for the original author, but it seems this was just the way to skip unused entries in the array. As the current approach is not compliant with the RFC, I've changed this to check if the reply callback is set instead.